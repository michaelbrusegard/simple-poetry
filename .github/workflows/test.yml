name: Test

on:
  push:
    branches:
      - main

jobs:
  test:
    name: Run Tests
    runs-on: self-hosted
    steps:
      - name: Checkout code
        uses: actions/checkout@v4

      - name: Set up Node.js
        uses: actions/setup-node@v4
        with:
          node-version: '22'

      - name: Install dependencies
        run: npm install

      - name: Run tests
        run: npx vitest run
