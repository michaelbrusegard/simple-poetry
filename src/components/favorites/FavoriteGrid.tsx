import styles from '../home/PoemsGrid.module.css'
import { PoemCard } from '../home/PoemCard'
import { Poem } from '../../types'
import { useEffect, useState } from 'react'

function FavoriteGrid() {
  const [favorites, setFavorites] = useState<Poem[]>([])

  useEffect(() => {
    const storedFavorites = localStorage.getItem('favorites')
    if (storedFavorites) {
      setFavorites(JSON.parse(storedFavorites))
    }
  }, [])

  return (
    <div className={styles.grid}>
      {favorites.length > 0 ? (
        favorites.map((poem: Poem) => <PoemCard key={poem.title} poem={poem} />)
      ) : (
        <p className={styles.errorMessage}>You don't have any favorites.</p>
      )}
    </div>
  )
}

export { FavoriteGrid }
