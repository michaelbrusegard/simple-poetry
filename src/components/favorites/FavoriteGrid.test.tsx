import { screen, act } from '@testing-library/react'
import { render } from '../../test/utils'
import { FavoriteGrid } from './FavoriteGrid'

describe('FavoriteGrid', () => {
  afterEach(() => {
    localStorage.clear()
  })

  it('should render the favorites from localStorage', () => {
    const favorites = [
      {
        title: 'The Chimney Sweeper',
        author: 'William Blake',
        linecount: '12',
      },
      {
        title: 'A Wind that rose',
        author: 'Emily Dickinson',
        linecount: '9',
      },
    ]
    localStorage.setItem('favorites', JSON.stringify(favorites))

    render(<FavoriteGrid />)

    expect(screen.getByText('The Chimney Sweeper')).toBeInTheDocument()
    expect(screen.getByText('William Blake')).toBeInTheDocument()
    expect(screen.getByText('12 lines')).toBeInTheDocument()
    expect(screen.getByText('A Wind that rose')).toBeInTheDocument()
    expect(screen.getByText('Emily Dickinson')).toBeInTheDocument()
    expect(screen.getByText('9 lines')).toBeInTheDocument()
  })

  it('should render a message if there are no favorites', () => {
    render(<FavoriteGrid />)

    expect(
      screen.getByText("You don't have any favorites."),
    ).toBeInTheDocument()
  })

  it('should match the snapshot', async () => {
    let fragment
    await act(async () => {
      const { asFragment } = render(<FavoriteGrid />)
      fragment = asFragment()
    })
    expect(fragment).toMatchSnapshot()
  })
})
