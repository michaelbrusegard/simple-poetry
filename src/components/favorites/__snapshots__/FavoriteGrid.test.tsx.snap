// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FavoriteGrid > should match the snapshot 1`] = `<DocumentFragment />`;
